contracts/KyberOasisReserve.sol

```diff
@@ -0,0 +1,258 @@
+
+
+import "./ERC20Interface.sol";
+import "./Utils.sol";
+import "./Withdrawable.sol";
+import "./KyberReserveInterface.sol";
+
+
+contract OtcInterface {
+    function getBuyAmount(address, address, uint) public constant returns (uint);
+}
+
+
+contract OasisDirectInterface {
+    function sellAllAmountPayEth(OtcInterface, ERC20, ERC20, uint)public payable returns (uint);
+    function sellAllAmountBuyEth(OtcInterface, ERC20, uint, ERC20, uint) public returns (uint);
+}
+
+
+contract KyberOasisReserve is KyberReserveInterface, Withdrawable, Utils {
+
+    address public kyberNetwork;
+    OasisDirectInterface public oasisDirect;
+    OtcInterface public otc;
+    ERC20 public wethToken;
+    ERC20 public tradeToken;
+    bool public tradeEnabled;
+    uint public feeBps;
+
+    function KyberOasisReserve(
+        address _kyberNetwork,
+        OasisDirectInterface _oasisDirect,
+        OtcInterface _otc,
+        ERC20 _wethToken,
+        ERC20 _tradeToken,
+        address _admin,
+        uint _feeBps
+    ) public {
+        require(_admin != address(0));
+        require(_oasisDirect != address(0));
+        require(_kyberNetwork != address(0));
+        require(_otc != address(0));
+        require(_wethToken != address(0));
+        require(_tradeToken != address(0));
+        require(_feeBps < 10000);
+
+        kyberNetwork = _kyberNetwork;
+        oasisDirect = _oasisDirect;
+        otc = _otc;
+        wethToken = _wethToken;
+        tradeToken = _tradeToken;
+        admin = _admin;
+        feeBps = _feeBps;
+        tradeEnabled = true;
+    }
+
+    function() public payable {
+        DepositToken(ETH_TOKEN_ADDRESS, msg.value);
+    }
+
+    event TradeExecute(
+        address indexed origin,
+        address src,
+        uint srcAmount,
+        address destToken,
+        uint destAmount,
+        address destAddress
+    );
+
+    function trade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        public
+        payable
+        returns(bool)
+    {
+
+        require(tradeEnabled);
+        require(msg.sender == kyberNetwork);
+
+        require(doTrade(srcToken, srcAmount, destToken, destAddress, conversionRate, validate));
+
+        return true;
+    }
+
+    event TradeEnabled(bool enable);
+
+    function enableTrade() public onlyAdmin returns(bool) {
+        tradeEnabled = true;
+        TradeEnabled(true);
+
+        return true;
+    }
+
+    function disableTrade() public onlyAlerter returns(bool) {
+        tradeEnabled = false;
+        TradeEnabled(false);
+
+        return true;
+    }
+
+    event ReserveParamsSet(
+        address kyberNetwork,
+        OasisDirectInterface oasisDirect,
+        OtcInterface otc,
+        ERC20 wethToken,
+        ERC20 tradeToken,
+        uint feeBps
+    );
+
+    function setReserveParams(
+        address _kyberNetwork,
```

Comment:
setting reserve params should be separated to single parameter functions. it will be easier to set and can help avoid set issues.
in Kyber reserve its one function since we tried to save gas, but that actually is bad.

Reply:
done.